### PR TITLE
Add support for collection of pipeline logs

### PR DESCRIPTION
**Pull request by FlorianVeaux**
### What does this PR do?

This PR adds the ability to forward logs emitted by Jenkins Pipelines directly to Datadog.

### Description of the Change

This PR adds a TaskListenerDecorator and the relevant tests.

### CODE REVIEW

File: src/main/java/org/datadog/jenkins/plugins/datadog/logs/DatadogDecorator.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import javax.annotation.Nullable;
+import org.datadog.jenkins.plugins.datadog.DatadogUtilities;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.jenkinsci.plugins.workflow.log.TaskListenerDecorator;
+
+public class DatadogDecorator extends TaskListenerDecorator {
+    private static final long serialVersionUID = 1L;
+    private static final Logger LOGGER = Logger.getLogger(DatadogDecorator.class.getName());
+    private transient Run<?, ?> run;
```
Comment (hithwen) [inline src/main/java/org/datadog/jenkins/plugins/datadog/logs/DatadogDecorator.java:44]: Why not `WorkflowRun` ?

File: src/test/java/org/datadog/jenkins/plugins/datadog/logs/LogCollectionTest.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    public void setup()
+    {
+        PowerMockito.mockStatic(ClientFactory.class);
+        when(ClientFactory.getClient()).thenReturn(stubClient);
+        DatadogGlobalConfiguration cfg = ExtensionList.lookup(DatadogGlobalConfiguration.class).get(0);
+        cfg.setCollectBuildLogs(true);
+    }
+
+    @Test
+    public void testFreeStyleProject() throws Exception {
```
Comment (sarah-witt) [inline src/test/java/org/datadog/jenkins/plugins/datadog/logs/LogCollectionTest.java:51]: Could you add a test case where logs are collected on another node?

File: src/test/java/org/datadog/jenkins/plugins/datadog/clients/DatadogClientStub.java [hunk 3]
```diff
@@ -36,12 +38,14 @@ of this software and associated documentation files (the "Software"), to deal
 
 public class DatadogClientStub implements DatadogClient {
 
-    public List<DatadogMetric> metrics = new ArrayList<>();
-    public List<DatadogMetric> serviceChecks = new ArrayList<>();
+    public List<DatadogMetric> metrics;
```
Comment (sarah-witt) [inline src/test/java/org/datadog/jenkins/plugins/datadog/clients/DatadogClientStub.java:41]: 👍

File: src/main/java/org/datadog/jenkins/plugins/datadog/clients/ClientFactory.java [hunk 4]
```diff
@@ -31,9 +31,18 @@ of this software and associated documentation files (the "Software"), to deal
 import org.datadog.jenkins.plugins.datadog.DatadogUtilities;
 
 public class ClientFactory {
+    private static DatadogClient testClient;
```
Comment (sarah-witt) [inline src/main/java/org/datadog/jenkins/plugins/datadog/clients/ClientFactory.java:34]: I like this idea, and how it eliminates use for PowerMock or lots of stubbing. I think if we want to unify the testing approach, we should make a separate PR on master with this, that way we can use it on all PRs. What do you think?

File: src/test/java/org/datadog/jenkins/plugins/datadog/logs/LogCollectionTest.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    @Test
+    public void testPipeline() throws Exception {
+        WorkflowJob p = j.jenkins.createProject(WorkflowJob.class, "pipeline");
+        p.setDefinition(new CpsFlowDefinition("echo 'foo'\n", true));
+        p.scheduleBuild2(0).get();
+        Assert.assertTrue(stubClient.logLines.contains("foo"));
+    }
+
+    @Test
+    public void testPipelineOnSlave() throws Exception {
```
Comment (sarah-witt) [inline src/test/java/org/datadog/jenkins/plugins/datadog/logs/LogCollectionTest.java:71]:
```suggestion
    public void testPipelineOnWorkerNode() throws Exception {
```

File: src/test/java/org/datadog/jenkins/plugins/datadog/logs/LogCollectionTest.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        WorkflowJob p = j.jenkins.createProject(WorkflowJob.class, "pipelineOnSlave");
```
Comment (sarah-witt) [inline src/test/java/org/datadog/jenkins/plugins/datadog/logs/LogCollectionTest.java:72]:
```suggestion
        WorkflowJob p = j.jenkins.createProject(WorkflowJob.class, "pipelineOnWorkerNode");
```

File: src/main/java/org/datadog/jenkins/plugins/datadog/clients/ClientFactory.java (hunk 4 above)
Comment (FlorianVeaux) [reply src/main/java/org/datadog/jenkins/plugins/datadog/clients/ClientFactory.java:34]: Yeah let's do that on another PR

File: src/test/java/org/datadog/jenkins/plugins/datadog/logs/LogCollectionTest.java (hunk 5 above)
Comment (FlorianVeaux) [reply src/test/java/org/datadog/jenkins/plugins/datadog/logs/LogCollectionTest.java:71]: why ? :)

Comment (sarah-witt) [reply src/test/java/org/datadog/jenkins/plugins/datadog/logs/LogCollectionTest.java:71]: The naming is not great and they are moving away from using this terminology https://www.jenkins.io/blog/2020/05/06/docker-agent-image-renaming/ https://issues.jenkins-ci.org/browse/JENKINS-42816
